### PR TITLE
[WIP] OSRA-299 Atomic sponsorship inactivation

### PR DESCRIPTION
**Pull request by NikitaAvvakumov**
https://osraav.atlassian.net/browse/OSRA-299

All changes associated with sponsorship inactivation wrapped in an AR
transaction to ensure that either all go through or none:

 - `sponsorship.update_attributes!`, `set_orphan_status_to_previously_sponsored`
   and `update_sponsor` are called from within `ActiveRecord::Base.transaction`
   block in `Sponsorship#inactivate`
 - any exceptions raised by the above methods are rescued by adding an error
   message to the sponsorship object and returning `false` to the controller
 - `set_orphan_status_to_sponsored` & `update_sponsor` moved to `after_create`

### CODE REVIEW

**Comment by NikitaAvvakumov** [comment]
Since `save` & `destroy` are wrapped in AR transactions by default, an alternative way to handle this would be to call sponsor & orphan update methods from the `after_save` hook. `update_sponsor` is already intelligent enough for this. However, the orphan update is currently split into two methods - `set_orphan_status_to_sponsored` & `set_orphan_status_to_previously_sponsored`. If sponsorship were to simply call `update_orphan` instead, and the Orphan model then figured which status it needs to set, then we could do the following:

```ruby
# app/models/sponsorship.rb

# after_create :update_sponsor, :set_orphan_status_to_sponsored
after_save :update_sponsor, :update_orphan

def inactivate(end_date)
  update_attributes!(active: false, end_date: end_date)
end
```
This implementation looks cleaner to me _and_ transfers the logic and responsibility from Sponsorship to Orphan where I think it belongs.

Thoughts?

File: spec/models/sponsorship_spec.rb [hunk 1]
```diff
@@ -128,6 +128,47 @@
...
       end
+
+      describe 'atomicity' do
+        let(:sponsor) { create :sponsor }
+        let(:orphan) { create :orphan }
+        let!(:sponsorship) do
+          create :sponsorship, sponsor: sponsor, orphan: orphan
+        end
+
+        def relevant_attributes
```
Comment (PurityControl) [inline spec/models/sponsorship_spec.rb]: why a method rather than a let?

Comment (NikitaAvvakumov) [inline spec/models/sponsorship_spec.rb]: Because `let` memoizes while the method resamples

**Comment by PurityControl** [comment]
This brings up the issue of refactoring Orphan. 

We have talked about tidying up father and how to tease out databases columns that are part of the table to do this. 

However there is another substantial part of the class that deals with sponsorship issues that orphan shouldn't have to know or worry about. 

I think we are missing an additional class here that has been left out of the domain say OrphanSponsorship for lack of imagining any better description at the moment. This additional class would be ideally suited to taking some of the sponsorship weight off Orphan's shoulders.

But I do not think sponsorship logic should be passed on to either the orphan or sponsor models.

**Comment by NikitaAvvakumov** [comment]
Extract a service object that handles sponsorship-related state changes for orphan & sponsor, and implement the observer pattern whereby sponsorship's state change notifies it to execute?
(It's very cool that I actually understand what I just said!)

I would very much like to combine or precede this with converting orphan's statuses to enums for state-machine-like behaviour.

**Comment by NikitaAvvakumov** [comment]
Will wait until model refactoring begins

**Pull request closed without merging**